src/community/label_propagation.c

```diff
@@ -213,163 +636,37 @@ igraph_error_t igraph_community_label_propagation(const igraph_t *graph,
-                    }
-                    VECTOR(label_counters)[k]++;
-                    if (VECTOR(label_counters)[k] == 1) {
-                        /* counter just became nonzero */
-                        IGRAPH_CHECK(igraph_vector_int_push_back(&nonzero_labels, k));
-                    }
-                    if (max_count < VECTOR(label_counters)[k]) {
-                        max_count = VECTOR(label_counters)[k];
-                        IGRAPH_CHECK(igraph_vector_int_resize(&dominant_labels, 1));
-                        VECTOR(dominant_labels)[0] = k;
-                    } else if (max_count == VECTOR(label_counters)[k]) {
-                        IGRAPH_CHECK(igraph_vector_int_push_back(&dominant_labels, k));
-                    }
-                }
-            }
+    switch (lpa_variant) {
+    case IGRAPH_LPA_FAST:
+        IGRAPH_CHECK(igraph_i_community_fast_label_propagation(graph, membership, mode, weights, fixed_copy));
+        break;
 
-            if (igraph_vector_int_size(&dominant_labels) > 0) {
-                if (control_iteration) {
-                    /* Check if the _current_ label of the node is also dominant */
-                    if (VECTOR(label_counters)[VECTOR(*membership)[v1]] != max_count) {
-                        /* Nope, we need at least one more iteration */
-                        running = true;
-                    }
-                }
-                else {
-                    /* Select randomly from the dominant labels */
-                    k = RNG_INTEGER(0, igraph_vector_int_size(&dominant_labels) - 1);
-                    VECTOR(*membership)[v1] = VECTOR(dominant_labels)[k];
-                }
-            }
+    case IGRAPH_LPA_RETENTION:
+        IGRAPH_CHECK(igraph_i_community_label_propagation(graph, membership, mode, weights, fixed_copy, /* retention */ true ));
+        break;
 
-            /* Clear the nonzero elements in label_counters */
-            num_neis = igraph_vector_int_size(&nonzero_labels);
-            for (j = 0; j < num_neis; j++) {
-                VECTOR(label_counters)[VECTOR(nonzero_labels)[j]] = 0;
-            }
-        }
-        RNG_END();
+    case IGRAPH_LPA_DOMINANCE:
+        IGRAPH_CHECK(igraph_i_community_label_propagation(graph, membership, mode, weights, fixed_copy, /* retention */ false));
+        break;
 
-        /* Alternating between control iterations and label updating iterations */
-        control_iteration = !control_iteration;
+    default:
+        IGRAPH_ERROR("Invalid igraph_lpa_variant_t.", IGRAPH_EINVAL);
     }
 
-    if (weights) {
-        igraph_inclist_destroy(&il);
-    } else {
-        igraph_adjlist_destroy(&al);
-    }
-    IGRAPH_FINALLY_CLEAN(1);
 
-    /* Shift back the membership vector, permute labels in increasing order */
-    /* We recycle label_counters here :) and use it as an integer vector from now on */
+    /* Permute labels in increasing order */
```

Comment:
Can you please restore the comment that explains that `label_counters` is used for an unrelated purpose here?

```
/* We recycle label_counters here :) and use it as an integer vector from now on */
```

It's quite confusing otherwise.

In fact it might be better to just eliminate any re-use and create a separate integer vector for this purpose.

Reply:
I've remove the `label_counters` vector, it wasn't used anywhere, and replaced it by an int vector.